Updated JsonRpc Api to v0.3.0-rc1

bartekryba opened the pull request:
Updated JsonRpc api to support [0.3.0-rc1](https://github.com/starkware-libs/starknet-specs/releases/tag/v0.3.0-rc1) spec.

## Usage related changes

<!-- How the changes from this PR affect users. -->
- Added Cairo1 class support
- Added support for declare v2
- It's now possible to estimate fee for multiple transactions
- Felt validation got updated (No more 0x0 prefixed felts)

## Development related changes

<!-- How these changes affect the developers of this project - e.g. changes in testing or CI/CD. -->

- For `run_devnet_in_background` method, the flag `--cairo-compiler-manifest` will be set by default if it wasn't explicitly set.

## Checklist:

- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/lint.sh`
- [x] Performed code self-review
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Documented the changes
- [ ] Linked the issues which this PR resolves
- [x] Updated the tests
- [x] All tests are passing - `./scripts/test.sh`

bartekryba committed:
Added cairo manifest flag, when running devnet from tests

bartekryba committed:
Updated rpc_spec file

bartekryba committed:
Updated getStateUpdate to 0.3

bartekryba committed:
Replaced 0x0 prefix with 0x

bartekryba committed:
Added new version of declare transaction

bartekryba committed:
ran format

bartekryba committed:
Applied lint suggestions

bartekryba committed:
Updated rpc spec

bartekryba committed:
Fixed test_rpc_call

bartekryba committed:
Fixed storage tests

bartekryba committed:
Updated estimateFee to support multiple transactions

bartekryba committed:
Removed adding deploy transaction from devnet

bartekryba committed:
Updated getEvents filter

bartekryba committed:
Merged master

bartekryba committed:
Fixed remaining tests

bartekryba committed:
Ran fomatter and linter

bartekryba committed:
Added test for get_class_at

bartekryba committed:
Updated get class tests

bartekryba committed:
Updated rpc docs

drknzz commented on page/docs/guide/json-rpc-api.md:
```diff
@@ -1,10 +1,10 @@
----
+--
```
Doesn't this break the file?

drknzz commented on starknet_devnet/blueprints/rpc/utils.py:
```diff
@@ -75,15 +75,25 @@ async def assert_block_id_is_valid(block_id: BlockId) -> None:
-        return "0x00"
-    return "0x0" + hex(value).lstrip("0x")
+        return "0x0"
+    return "0x" + hex(value).lstrip("0x")
```
Isn't this just `hex(value)` now?

bartekryba replied:
yeah 😅

drknzz commented on starknet_devnet/blueprints/rpc/utils.py:
```diff
@@ -101,7 +111,7 @@ def rpc_root(root: str) -> Felt:
     Convert 0 prefixed root to 0x prefixed root
     """
-    return "0x0" + (root.lstrip("0") or "0")
+    return "0x" + (root.lstrip("0") or "0")
```
Could `rpc_felt` be used here? It seems that root is defined as felt in the spec

drknzz commented on test/rpc/rpc_utils.py:
```diff
@@ -115,4 +115,4 @@ def is_felt(value: str) -> bool:
     Check whether value is a Felt
     """
-    return bool(re.match(r"^0x0[a-fA-F0-9]{1,63}$", value))
+    return bool(re.match(r"^0x[a-fA-F0-9]{1,63}$", value))
```
The felt regex seems to have been changed
```
"pattern": "^0x(0|[a-fA-F1-9]{1}[a-fA-F0-9]{0,62})$"
```

drknzz commented on test/rpc/test_data/get_events.py:
```diff
@@ -45,7 +45,8 @@ def create_get_events_filter(
     """
     if keys is None:
-        keys = [rpc_felt(FEE_CHARGED_EVENT_KEY)]
+        keys = []
```
Was this omitted on purpose?

bartekryba replied:
Updated it

drknzz commented on starknet_devnet/blueprints/rpc/misc.py:
```diff
@@ -41,11 +41,20 @@ def check_address(address, event):
     """
-    return bool(set(event.keys) & set(keys))
+    if not keys:
+        return True
```
Wouldn't the below code just return `True` if keys were empty?

drknzz replied:
Or can keys be `None`? If so, let's maybe explicitly check `if keys is None` and also add `Optional` to the typing

drknzz commented on starknet_devnet/blueprints/rpc/misc.py:
```diff
@@ -139,7 +148,11 @@ async def get_events(
+    if filter.get("keys"):
+        keys = [[int(key, 0) for key in keys] for keys in filter.get("keys")]
+    else:
+        keys = None
```
Similar situation here, the `else` branch would execute in both cases when `keys` are not specified, and when they are an empty list.

bartekryba committed:
Apply code review suggestions

bartekryba committed:
Merge branch 'master' into rpc-v3

bartekryba committed:
Updated events filter

bartekryba committed:
Merge branch 'master' into rpc-v3

bartekryba committed:
Updated check_keys body

bartekryba committed:
Updated check_keys comment

bartekryba committed:
Made check_keys method private.

bartekryba committed:
Made is_felt method use pattern from rpc_spec file.

bartekryba committed:
Replaced inline test resources with artifacts.

Replaced inline test resources with compiled contract artifacts,
formatted some files.

bartekryba committed:
Created constants for declare tx versions.

bartekryba committed:
Removed support for declare v0 transaction

bartekryba committed:
Removed unnecessary deploy from getClass test.

bartekryba committed:
Updated getStateUpdate to return pending state.

bartekryba committed:
Updated test_get_pending_state_update
